doc/src/extras.rst

```diff
@@ -141,6 +141,262 @@ Logging cursor
+    `~start_replication()` handles all the complexity, but at times it might
+    be beneficial to use low-level interface for better control, in particular
+    to `~select.select()` on multiple sockets.  The following methods are
+    provided for asynchronous operation:
+
+    .. method:: read_replication_message(decode=True)
+
+        :param decode: a flag indicating that unicode conversion should be
+                       performed on the data received from the server
+
+        This method should be used in a loop with asynchronous connections
+        after calling `~start_replication()` once.
+
+        It tries to read the next message from the server, without blocking
+        and returns an instance of `~ReplicationMessage` or *None*, in case
+        there are no more data messages from the server at the moment.
+
+        It is expected that the calling code will call this method repeatedly
+        in order to consume all of the messages that might have been buffered,
+        until *None* is returned.  After receiving a *None* value from this
+        method, one might use `~select.select()` or `~select.poll()` on the
+        corresponding connection to block the process until there is more data
+        from the server.
+
+        The server can send keepalive messages to the client periodically.
+        Such messages are silently consumed by this method and are never
+        reported to the caller.
+
+    .. method:: fileno()
+
+        Call the corresponding connection's `~connection.fileno()` method and
+        return the result.
+
+        This is a convenience method which allows replication cursor to be
+        used directly in `~select.select()` or `~select.poll()` calls.
+
+    .. attribute:: replication_io_timestamp
+
+        A `~datetime` object representing the timestamp at the moment of last
+        communication with the server (a data or keepalive message in either
+        direction).
+
+    An actual example of asynchronous operation might look like this::
+
+        keepalive_interval = 10.0
+        while True:
+            if (datetime.now() - cur.replication_io_timestamp).total_seconds() >= keepalive_interval:
+                cur.send_replication_feedback()
+
+            while True:
+                msg = cur.read_replication_message()
+                if not msg:
+                    break
+                writer.write(msg)
+
+            timeout = keepalive_interval - (datetime.now() - cur.replication_io_timestamp).total_seconds()
+            if timeout > 0:
+                select.select([cur], [], [], timeout)
+
+.. autoclass:: ReplicationMessage
+
+    .. attribute:: payload
+
+        The actual data received from the server.  An instance of either
+        ``str`` or ``unicode``.
+
+    .. attribute:: data_start
+
+        LSN position of the start of the message.
+
+    .. attribute:: wal_end
+
+        LSN position of the end of the message.
```

Comment:
I think you've misread the docs there.

See http://www.postgresql.org/docs/9.4/static/protocol-replication.html, message XLogData (B)

> Int64 The current end of WAL on the server.

The docs indicate that this message field is the *server's current xlog end position*, not the xlog position for the end of the message. Otherwise it'd be redundant anyway, since one message's end would be the next one's start.